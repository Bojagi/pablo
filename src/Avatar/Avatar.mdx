---
name: Avatar
route: /avatar
menu: Components
---

import { Props, Playground } from 'docz'
import { Avatar } from './Avatar.tsx'
import { Flex } from '../Box/Box.tsx';

# Avatar

Renders an image with square dimensions and masks it with a circle or rounded corners.

## Examples

### Different sizes

Avatars can have different sizes, spaning from `tiny` to `large`. By default avatars are square.

<Playground>
  <Flex mx={-4}>
    <Avatar
      mx={4}
      size="large"
      src="https://avatars.githubusercontent.com/u/36902682?s=200&v=4"
      variant="square"
    />
    <Avatar
      size="medium"
      mx={4}
      src="https://avatars.githubusercontent.com/u/36902682?s=200&v=4"
      variant="square"
    />
    <Avatar
      size="small"
      mx={4}
      src="https://avatars.githubusercontent.com/u/36902682?s=200&v=4"
      variant="square"
    />
    <Avatar
      size="tiny"
      mx={4}
      src="https://avatars.githubusercontent.com/u/36902682?s=200&v=4"
      variant="square"
    />
  </Flex>
</Playground>

### Circle
<Playground>
  <Avatar
    size="large"
    src="https://avatars.githubusercontent.com/u/36902682?s=200&v=4"
    variant="circle"
  />
</Playground>

## Import

```js
import { Avatar } from '@bojagi/pablo/Avatar';
```

## Props
<Props of={Avatar} />
