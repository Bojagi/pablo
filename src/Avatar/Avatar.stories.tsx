import React from 'react';
import { css } from '@emotion/react';
import { Flex } from '../Box';
import { Avatar } from './index';

export default {
  title: 'Avatar',
};

const BaseStory = (args) => (
  <Flex mx={-1}>
    <Avatar
      mx={1}
      size="large"
      src="https://avatars.githubusercontent.com/u/36902682?s=200&v=4"
      {...args}
    />
    <Avatar
      size="medium"
      mx={1}
      src="https://avatars.githubusercontent.com/u/36902682?s=200&v=4"
      {...args}
    />
    <Avatar
      size="small"
      mx={1}
      src="https://avatars.githubusercontent.com/u/36902682?s=200&v=4"
      {...args}
    />
    <Avatar
      size="tiny"
      mx={1}
      src="https://avatars.githubusercontent.com/u/36902682?s=200&v=4"
      {...args}
    />
  </Flex>
);

export const Square = BaseStory.bind(null);
Square.args = {
  variant: 'square',
};

export const Circle = BaseStory.bind(null);
Circle.args = {
  variant: 'circle',
};

export const CustomStyleCircle = BaseStory.bind(null);
CustomStyleCircle.args = {
  variant: 'circle',
  customStyles: {
    circle: css`
      border: 4px solid white;
      box-shadow: 3px 3px 3px rgba(0, 0, 0, 0.2);
    `,
  },
};
