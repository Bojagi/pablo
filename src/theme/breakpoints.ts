export type Breakpoint = 'sm' | 'md' | 'lg' | 'xl';

export const breakpointNames: Breakpoint[] = ['sm', 'md', 'lg', 'xl'];

export type BreakpointsArray = [number, number, number, number];

export type Breakpoints = BreakpointsArray &
  Record<Breakpoint, number> & {
    breakpointNames: Breakpoint[];
  };

export function createBreakpoints(
  array: BreakpointsArray,
  usedBreakpointNames: Breakpoint[] = breakpointNames
): Breakpoints {
  const newBreakpoints = ([...array] as unknown) as Breakpoints;
  usedBreakpointNames.forEach((bp, index) => {
    newBreakpoints[bp] = newBreakpoints[index];
  });
  newBreakpoints.breakpointNames = usedBreakpointNames;

  return newBreakpoints as Breakpoints;
}

export const defaultBreakpoints: BreakpointsArray = [600, 1000, 1200, 1920];

export const breakpoints = createBreakpoints(defaultBreakpoints, breakpointNames);
