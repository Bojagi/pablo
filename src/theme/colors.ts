export type ShadedColor = {
  main: string;
  dark: string;
  darkest: string;
  light: string;
  lightest: string;
  contrastText: string;
  contrastTextLight: string;
};

export type CommonColors = {
  white: string;
  black: string;
  whiteContrastText: string;
  blackContrastText: string;
};

export type TextColors = {
  main: string;
  info: string;
};

export type BorderColors = {
  main: string;
  lightest: string;
};

export interface GrayColors {
  '50': string;
  '100': string;
  '200': string;
  '300': string;
  '400': string;
  '500': string;
  '600': string;
  '700': string;
  '800': string;
  '900': string;
}

export type AllColors = ShadedColor & TextColors & BorderColors & CommonColors;

export interface Colors {
  common: CommonColors;
  gray: GrayColors;
  borders: BorderColors;
  brand: ShadedColor;
  positive: ShadedColor;
  negative: ShadedColor;
  neutral: ShadedColor;
  text: TextColors;
}

export const WHITE = '#ffffff';
export const BLACK = '#000000';

export const colors: Colors = {
  common: {
    white: WHITE,
    black: BLACK,
    whiteContrastText: BLACK,
    blackContrastText: WHITE,
  },
  gray: {
    '50': getGrayColor(0.95),
    '100': getGrayColor(0.9),
    '200': getGrayColor(0.8),
    '300': getGrayColor(0.7),
    '400': getGrayColor(0.6),
    '500': getGrayColor(0.5),
    '600': getGrayColor(0.4),
    '700': getGrayColor(0.3),
    '800': getGrayColor(0.2),
    '900': getGrayColor(0.1),
  },
  borders: {
    main: `${BLACK}${getHexByte(0.25)}`,
    lightest: `${BLACK}${getHexByte(0.1)}`,
  },
  text: {
    main: BLACK,
    info: `${BLACK}${getHexByte(0.5)}`,
  },
  brand: {
    lightest: '#EBE7FF',
    light: '#D6CEFF',
    main: '#6A50F2',
    dark: '#4129BD',
    darkest: '#281683',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  positive: {
    lightest: '#DFF6E6',
    light: '#BBECCB',
    main: '#38C765',
    dark: '#2FB85A',
    darkest: '#108B36',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  negative: {
    lightest: '#FFE7E5',
    light: '#FFCFCC',
    main: '#F21D0D',
    dark: '#CA271B',
    darkest: '#871A12',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
  neutral: {
    lightest: '#FCF5DB',
    light: '#FFEB9E',
    main: '#FFDD5B',
    dark: '#EDC52A',
    darkest: '#D6AC0B',
    contrastText: WHITE,
    contrastTextLight: BLACK,
  },
};

/**
 * Get hex color byte (e.g. to get transparency)
 * @param value Percentage value (between 0.0 and 1.0)
 */
function getHexByte(value: number) {
  return (Math.round((255 * value) / 16) * 16).toString(16);
}

function getGrayColor(value: number) {
  return `#${getHexByte(value)}${getHexByte(value)}${getHexByte(value)}`;
}
