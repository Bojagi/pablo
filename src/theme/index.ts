export * from './defaultTheme';
export * from './context';
export * from './PabloThemeProvider';
export * from './useComponentStyle';
