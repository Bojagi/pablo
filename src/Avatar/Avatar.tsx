import React, { forwardRef } from 'react';
import styled from '@emotion/styled';
import { css } from '@emotion/react';
import { Image } from '../Image';
import { useComponentStyle } from '../theme/useComponentStyle';
import { conditionalStyles, getComponentStyle } from '../styleHelpers';
import { useCustomStyles } from '../utils/useCustomStyles';
import { baseStyle } from '../shared/baseStyle';
import type { FC, RefAttributes, ForwardRefExoticComponent } from 'react';
import type { LayoutBoxProps } from '../Box';
import type { ImageProps } from '../Image';
import type { BaseProps } from '../types';
import type { AvatarStyleProperties } from './styles';

export type AvatarSize = 'tiny' | 'small' | 'medium' | 'large';
export type AvatarVariant = 'square' | 'circle';

export interface AvatarProps extends LayoutBoxProps, BaseProps<AvatarStyleProperties> {
  className?: string;
  /**
   * URL to the image you want to use as an avatar
   */
  src: string;
  /**
   * Size of the avatar
   */
  size?: AvatarSize;
  /**
   * Define if you want the avatar to be circle or square shaped
   */
  variant?: AvatarVariant;
}

interface AvatarImageProps extends ImageProps {
  size: AvatarSize;
  variant: AvatarVariant;
}

const AvatarImage = styled<
  ForwardRefExoticComponent<AvatarImageProps & RefAttributes<HTMLImageElement>>
>(Image as any)`
  ${baseStyle}
  ${conditionalStyles('variant', (props) => ({
    square: css`
      border-radius: ${getComponentStyle('avatar.square.borderRadius')(props)}px;
    `,
    circle: css`
      border-radius: 50%;
    `,
  })) as any}
`;

export const Avatar: FC<AvatarProps> = forwardRef<HTMLImageElement, AvatarProps>(
  ({ className, src, size = 'medium', variant = 'square', customStyles, ...props }, ref) => {
    const dimension = useComponentStyle('avatar.size.{size}', { size }) as number;
    const getCustomStyles = useCustomStyles('avatar.styles', customStyles);
    return (
      <AvatarImage
        ref={ref}
        {...props}
        variant={variant}
        size={size}
        width={dimension}
        height={dimension}
        css={getCustomStyles(variant)}
        className={className}
        src={src}
      />
    );
  }
);
