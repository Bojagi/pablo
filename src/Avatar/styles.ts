import { registerComponentStyles } from '../theme/defaultComponentStyles';
import { BaseStyles } from '../types';

export type AvatarStyleProperties = 'circle' | 'square';

export interface AvatarStyles extends BaseStyles<AvatarStyleProperties> {
  square: {
    borderRadius: number;
  };
  size: {
    tiny: number;
    small: number;
    medium: number;
    large: number;
  };
}
export const avatarStyles: AvatarStyles = {
  square: {
    borderRadius: 6,
  },
  size: {
    tiny: 24,
    small: 32,
    medium: 40,
    large: 64,
  },
};

registerComponentStyles('avatar', avatarStyles);
