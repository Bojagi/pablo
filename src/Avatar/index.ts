import './styles';

export * from './Avatar';
