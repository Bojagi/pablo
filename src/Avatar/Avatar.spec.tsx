import { render } from '@testing-library/react';
import React from 'react';
import { PabloThemeProvider } from '../theme';
import { Avatar } from './Avatar';

describe.each([
  [undefined, '6px'],
  ['square', '6px'],
  ['circle', '50%'],
])('%s variant', (variant, expectedBorderRadius) => {
  test.each([
    [undefined, 40],
    ['tiny', 24],
    ['small', 32],
    ['medium', 40],
    ['large', 64],
  ])('Render size %s', (size, extendedSize) => {
    const { container, getByTestId } = renderComponent({
      variant,
      size,
    });
    expect(container).toMatchSnapshot();
    expect(getByTestId('pbl-avatar')).toHaveStyleRule('width', `${extendedSize}px`);
    expect(getByTestId('pbl-avatar')).toHaveStyleRule('height', `${extendedSize}px`);
    expect(getByTestId('pbl-avatar')).toHaveStyleRule('height', `${extendedSize}px`);
    expect(getByTestId('pbl-avatar')).toHaveStyleRule('border-radius', expectedBorderRadius);
  });
});

function renderComponent(props) {
  return render(
    <PabloThemeProvider>
      <Avatar data-testid="pbl-avatar" {...props} />
    </PabloThemeProvider>
  );
}
