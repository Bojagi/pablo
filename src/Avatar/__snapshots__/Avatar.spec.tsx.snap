// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`circle variant Render size large 1`] = `
.c0 {
  box-sizing: border-box;
  width: 64px;
  height: 64px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 50%;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="64"
    variant="circle"
    width="64"
  />
</div>
`;

exports[`circle variant Render size medium 1`] = `
.c0 {
  box-sizing: border-box;
  width: 40px;
  height: 40px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 50%;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="40"
    variant="circle"
    width="40"
  />
</div>
`;

exports[`circle variant Render size small 1`] = `
.c0 {
  box-sizing: border-box;
  width: 32px;
  height: 32px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 50%;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="32"
    variant="circle"
    width="32"
  />
</div>
`;

exports[`circle variant Render size tiny 1`] = `
.c0 {
  box-sizing: border-box;
  width: 24px;
  height: 24px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 50%;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="24"
    variant="circle"
    width="24"
  />
</div>
`;

exports[`circle variant Render size undefined 1`] = `
.c0 {
  box-sizing: border-box;
  width: 40px;
  height: 40px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 50%;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="40"
    variant="circle"
    width="40"
  />
</div>
`;

exports[`square variant Render size large 1`] = `
.c0 {
  box-sizing: border-box;
  width: 64px;
  height: 64px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 6px;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="64"
    variant="square"
    width="64"
  />
</div>
`;

exports[`square variant Render size medium 1`] = `
.c0 {
  box-sizing: border-box;
  width: 40px;
  height: 40px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 6px;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="40"
    variant="square"
    width="40"
  />
</div>
`;

exports[`square variant Render size small 1`] = `
.c0 {
  box-sizing: border-box;
  width: 32px;
  height: 32px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 6px;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="32"
    variant="square"
    width="32"
  />
</div>
`;

exports[`square variant Render size tiny 1`] = `
.c0 {
  box-sizing: border-box;
  width: 24px;
  height: 24px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 6px;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="24"
    variant="square"
    width="24"
  />
</div>
`;

exports[`square variant Render size undefined 1`] = `
.c0 {
  box-sizing: border-box;
  width: 40px;
  height: 40px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 6px;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="40"
    variant="square"
    width="40"
  />
</div>
`;

exports[`undefined variant Render size large 1`] = `
.c0 {
  box-sizing: border-box;
  width: 64px;
  height: 64px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 6px;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="64"
    variant="square"
    width="64"
  />
</div>
`;

exports[`undefined variant Render size medium 1`] = `
.c0 {
  box-sizing: border-box;
  width: 40px;
  height: 40px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 6px;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="40"
    variant="square"
    width="40"
  />
</div>
`;

exports[`undefined variant Render size small 1`] = `
.c0 {
  box-sizing: border-box;
  width: 32px;
  height: 32px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 6px;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="32"
    variant="square"
    width="32"
  />
</div>
`;

exports[`undefined variant Render size tiny 1`] = `
.c0 {
  box-sizing: border-box;
  width: 24px;
  height: 24px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 6px;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="24"
    variant="square"
    width="24"
  />
</div>
`;

exports[`undefined variant Render size undefined 1`] = `
.c0 {
  box-sizing: border-box;
  width: 40px;
  height: 40px;
}

.c1 {
  box-sizing: border-box;
  border-radius: 6px;
}

<div>
  <img
    class="c0 c1"
    css=","
    data-testid="pbl-avatar"
    height="40"
    variant="square"
    width="40"
  />
</div>
`;
